k2/csrc/rnnt_decode.cu

```diff
@@ -823,55 +850,95 @@ void RnntDecodingStreams::FormatOutput(const std::vector<int32_t> &num_frames,
   *out_map = Array1<int32_t>(c_, num_arcs);
   int32_t *out_map_data = out_map->Data();
 
-  int32_t *oshape_row_ids3 = oshape.RowIds(3).Data(),
+
+  int32_t *oshape_row_ids4 = oshape.RowIds(4).Data(),
+          *oshape_row_ids3 = oshape.RowIds(3).Data(),
           *oshape_row_ids2 = oshape.RowIds(2).Data(),
           *oshape_row_ids1 = oshape.RowIds(1).Data(),
+          *oshape_row_splits3 = oshape.RowSplits(3).Data(),
           *oshape_row_splits2 = oshape.RowSplits(2).Data(),
           *oshape_row_splits1 = oshape.RowSplits(1).Data();
 
+  int32_t vocab_size = config_.vocab_size,
+          arc_map_b_num_elements = 0;
+  // t2s2c_shape is short for time2stream2context_shape
+  const int32_t *t2s2c_shape_row_splits2 = nullptr,
+                *t2s2c_shape_row_splits1 = nullptr;
+
+  // In `GatherPrevFrames`, we use
+  // Unstack(frames, 1, false /*pad_right*/, &prev_frames);
+  // to generated prev_frames.
+  // Since pad_right=false, shorter utterances are padded to the left side,
+  // i.e. padded beginning frames.
+  // We need to subtract number of padded frames when calculating
+  // the real time index.
+  auto num_padded_frames = Array1<int32_t>(c_, num_frames);
```

Comment:
You initialize `num_padded_frames` with real frames here, it takes me a while to get your idea till I read the following lines. Please add comments here.